### PR TITLE
Issue#616

### PR DESCRIPTION
**Pull request by Alapan**
Link - https://github.com/apinf/api-umbrella-dashboard/issues/616

### CODE REVIEW

File: client/views/api_backends/delete/deleteApiBackendConfirmation.html [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<template name="deleteApiBackendConfirmation">
+  <div class="modal fade">
+    <div class="modal-dialog">
+      <div class="modal-content">
+        <div class="modal-header">
+          <button type="button" class="close" data-dismiss="modal">&times;</button>
+          <h3 class="modal-title">{{_ "apiBackends_Delete_API_Title"}}</h3>
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.html]: Rename this i18n token to `deleteApiBackendConfirmation-header` and add corresponding translation string to en.i18n.js.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.html [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        </div>
+        <div class="modal-body">
+          <div id="confirmDelete">
+            <p>{{_ "delete_Message"}} {{ apiBackend }}? </p>
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.html]: Rename this i18n token to `deleteApiBackendConfirmation-confirmDeleteMessage` and add corresponding translation string to en.i18n.js.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.html [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            <p><span class="text-danger">{{_ "delete_Note"}} </span> {{_ "deleteInfo_Modal"}} </p>
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.html]: Rename this i18n token to `deleteApiBackendConfirmation_confirmDeleteNote` and add corresponding translation string to en.i18n.js.

What is the purpose of the `deleteInfo_Modal` token?

File: client/views/api_backends/delete/deleteApiBackendConfirmation.html [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+          </div>
+          {{#if restCallStarted }}
+            {{> spinner}}
+          {{/if}}
+          <div id="successDelete" class="hide">
+            <p>{{ savedBackendName }} {{_ "delete_Success"}}</p>
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.html]: Rename this i18n token to `deleteApiBackendConfirmation_successMessage` and add corresponding translation string to en.i18n.js.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.html [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+          </div>
+          <div id="failureDelete" class="hide">
+            <p><span class="text-danger">{{_ "delete_Error"}}</span> {{ savedBackendName }} {{_ "delete_Failure"}} </p>
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.html]: Rename this i18n token to `deleteApiBackendConfirmation_errorMessage` and add corresponding translation string to en.i18n.js.

Rename this i18n token to `deleteApiBackendConfirmation_failureMessage` and add corresponding translation string to en.i18n.js.

Do the error and failure messages need to be separate strings, or can they be combined?

File: client/views/api_backends/delete/deleteApiBackendConfirmation.html [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+          </div>
+        </div>
+        <div class="modal-footer">
+          <div id="confirmFooter">
+            <button type="button" class="btn btn-default" data-dismiss="modal">{{_ "view_ApiBackend_Details_cancelButtonText"}}</button>
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.html]: Rename this i18n token to `deleteApiBackendConfirmation_CancelButton` and add corresponding translation string to en.i18n.js.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.html [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            <button type="button" class="btn btn-danger" id="deleteApi">{{_ "view_ApiBackend_Details_deleteButtonText"}}</button>
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.html]: Rename this i18n token to `deleteApiBackendConfirmation_deleteButton` and add corresponding translation string to en.i18n.js.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.html [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+          </div>
+          <div id="doneFooter" class="hide">
+            <button type="button" class="btn btn-primary" data-dismiss="modal">{{_ "close"}}</button>
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.html]: Rename this i18n token to `deleteApiBackendConfirmation_closeButton` and add corresponding translation string to en.i18n.js.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
+Template.deleteApiBackendConfirmation.helpers({
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Template helpers are used to pass data in to templates, and don't typically have other side effects. What are the session variables used for? Consider declaring session variables in a different place, such as the onRendered callback or a button event.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  // save the API backend name and ID in session variables
+  'apiBackend': function() {
+    Session.set('apiBackendId', this._id);
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Consider declaring session variables in a different place, such as the onRendered callback or a button event. Determine whether the session variables could be replaced by template.instance variables.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    const apiBackendName = ApiBackends.findOne(this._id).name;
+    Session.set('apiBackendName', apiBackendName);
+    return apiBackendName;
+  },
+  // return saved API name to be displayed in success/error message.
+  // The API backend name is saved separately so that it can be 
+  // used even after the backend has been deleted.
+  'savedBackendName': function() {
+    return Session.get('apiBackendName');
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Use a template instance variable rather than a session variable, since this variable is used from within the same template in which it is declared.

E.g. 

```js
Template.deleteApiBackendConfirmation.onCreated(function () {
  // Get reference to template instance
  const instance = this;

 // Create instance variables
  instance.backendId = instance.data._id; 
  instance.backendName = instance.data.name;
});

Template.deleteApiBackendConfirmation.helpers({
  "savedBackendName": function () {
    // Get reference to template instance
    const instance = Template.instance();

    return instance.backendName;
  }
});
```

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  'click #deleteApi': function() {
+    const apiBackendDoc = ApiBackends.findOne(Session.get("apiBackendId"));
+    const apiUmbrellaApiId = apiBackendDoc.id;
+    const apiBackendId = Session.get("apiBackendId");
+
+    // Disable delete button to prevent multiple clicks
+    $('#deleteApi').prop("disabled", true);
+
+    // REST call started, start spinner
+    Session.set('restCallStarted', true);
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: What is this session variable used for? Consider using a template reactive variable instead. Basically, we aren't using session very much in this project, because there are usually ways to do things in the local scope.

**Comment by brylie** [comment]
Try not to use session variables much. There are usually ways to do things in the local scope. E.g. template instance variables and reactive variables.

**Comment by brylie** [comment]
All i18n tokens should be unique and prefixed with the template name. We are not re-using translations, as per the guidance of @baijat.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    // REST call to Admin API for deletion from Umbrella
+    Meteor.call('deleteApiBackendOnApiUmbrella', apiUmbrellaApiId, function(error, apiUmbrellaWebResponse) {
+
+      if (apiUmbrellaWebResponse.http_status === 204) {
+
+        // call method to remove API backend from collections
+        Meteor.call('removeApiBackend', apiBackendId);
+
+        $('#confirmDelete').hide(function() {
+          $('#successDelete').removeClass('hide');
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Consider using the s-alert package to display the success message. That way we don't have to manually handle showing and hiding the success dom element.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+  },
+
+  'restCallStarted': function() {
+    return Session.get('restCallStarted');
+  }
+});
+
+Template.deleteApiBackendConfirmation.events({
+  'click #deleteApi': function() {
+    const apiBackendDoc = ApiBackends.findOne(Session.get("apiBackendId"));
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Use instance variable(s) here rather than session variables.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+        if (currentRoute === 'viewApiBackend') {
+          Router.go('catalogue');
+        } else if (currentRoute === 'manageApiBackends') {
+          Router.go('manageApiBackends');
+        }
+
+      } else {
+        $('#confirmDelete').hide(function() {
+          $('#failureDelete').removeClass('hide');
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Use s-alert rather than a dom element here.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+        });
+      }
+      // REST call ended, stop spinner
+      Session.set('restCallStarted', false);
+      $('#confirmFooter').hide(function() {
+        $('#doneFooter').removeClass('hide');
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Use s-alert rather than a DOM element here.

File: client/views/api_backends/manage/manage.html [hunk 17]
```diff
@@ -16,6 +16,12 @@ <h2 class="panel-title">
              class="btn btn-warning">
             {{_ "manage_ApiBackends_Edit"}}
           </a>
+          <a
+             href="#"
+             class="btn btn-danger"
+             id="deleteModal">
+            {{_ "manage_ApiBackends_Delete"}}
```
Comment (brylie) [inline client/views/api_backends/manage/manage.html]: Rename this i18n token to `manageApiBackends_deleteButton` and add corresponding translation string to en.i18n.js.

File: client/views/api_backends/view_api_backend/details/details.html [hunk 18]
```diff
@@ -2,12 +2,19 @@
   <div class="panel panel-default">
     <div class="panel-heading">
       {{# if apiBackend.currentUserCanEdit }}
+        <a 
+           href="#"
+           class="btn btn-danger pull-right"
+           id="deleteModal">
+          {{_ "view_ApiBackend_Details_deleteButtonText"}}
```
Comment (brylie) [inline client/views/api_backends/view_api_backend/details/details.html]: I realize the surrounding i18n tokens will be inconsistent with this suggestion, but go ahead and rename this i18n token to `viewApiBackendDetails_deleteButton`. Then, add corresponding translation string to en.i18n.js.

**Comment by brylie** [comment]
Overall, great work here! This is a big feature to tackle.

Please review the above suggestions, and let me know if I may be of assistance.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.html (hunk 3 above)
Comment (Alapan) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.html]: The tokens in the <p> tags should have been written as 1, but I wanted to write 'NOTE' in red (span class='text-danger'), and the rest in normal black coloured text. However, I couldn't figure out how to write the span in the en.i18n.json file.  I changed 'deleteInfo_Modal'  to 'deleteApiBackendConfirmation_confirmDeleteInformation' though, if it helps.

**Comment by Alapan** [comment]
@brylie I made almost all the changes you wanted, only thing was I was not sure how to include `span class='text-danger'` in a token (explained in a comment earlier).

File: client/views/api_backends/delete/deleteApiBackendConfirmation.html (hunk 3 above)
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.html]: Alright, go ahead and leave the translation strings separate. However, the tokens still need the template name prefix, as mentioned previously.

`deleteApiBackendConfirmation-confirmDeleteNote`

**Comment by Alapan** [comment]
@brylie Another point was with s-alert, when deleting an API from the API details page (not the manage API page), the s-alert was not appearing when the deletion was successful and the user had been returned to the API list page (catalogue). I don't know if this is related to how fast the catalogue page is loading.

**Comment by brylie** [comment]
@Alapan, the s-alert is not high priority. It won't block this code from being merged. We can just open an enhancement, after finishing this task :-)

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+        
+        // based on name of current route, load suitable parent page
+        var currentRoute = Router.current().route.getName();
+
+        if (currentRoute === 'viewApiBackend') {
+          Router.go('catalogue');
+        } else if (currentRoute === 'manageApiBackends') {
+          Router.go('manageApiBackends');
+        }
+        sAlert.success(instance.backendName + " was successfully deleted!");
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: This may be out of scope. I.e. it may need to be moved in to one of the `if`, `else if`, or `else` blocks.

Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: This may be out of scope. I.e. it may need to be moved in to one of the `if`, `else if`, or `else` blocks.

Comment (Alapan) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: I found a hook "Router.onAfterAction", which shows the s-alert after the catalogue page is loading, but now it shows twice at the same time.

**Comment by brylie** [comment]
Try moving the sAlert call into the `if ... else if .. else` block. If that works great! If not lets remove it.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+      if (apiUmbrellaWebResponse.http_status === 204) {
+
+        // call method to remove API backend from collections
+        Meteor.call('removeApiBackend', instance.backendId);        
+        
+        // based on name of current route, load suitable parent page
+        var currentRoute = Router.current().route.getName();
+
+        if (currentRoute === 'viewApiBackend') {
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Rewrite this block using a [switch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch). This will improve readability and possibly reduce the probability of semantic errors.

Comment (Alapan) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: I thought about using switch-case earlier, but I wasn't sure what to use for the default case.

Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Perhaps just `break` in default.

Comment (Alapan) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Right now, I don't have anything for `else`, as I am not sure if I can set any default route. I can replace the `else if` with a separate `if` statement, though.

**Comment by Alapan** [comment]
@brylie I used switch-case and moved the s-alert inside each case, but the result was the same as before. I tried `Router.onAfterAction` when returning to the Catalogue page. It partially worked, but then the s-alert was showing twice.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+    $('#deleteApi').prop("disabled", true);
+
+    // start spinner when calling API
+    instance.restCallStarted.set(true);
+
+    // REST call to Admin API for deletion from Umbrella
+    Meteor.call('deleteApiBackendOnApiUmbrella', apiUmbrellaApiId, function(error, apiUmbrellaWebResponse) {
+
+      if (apiUmbrellaWebResponse.http_status === 204) {
+
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: As was pointed out in review, make sure to publish the changes on API Umbrella. This is a design of API Umbrella that changes do not take effect, at the nginx level, until they are published. We have some [example code](https://github.com/apinf/api-umbrella-dashboard/blob/develop/client/views/api_backends/add/wizard.js#L128-L129) you can review.

File: client/views/api_backends/view_api_backend/details/details.html [hunk 22]
```diff
@@ -2,12 +2,19 @@
   <div class="panel panel-default">
     <div class="panel-heading">
       {{# if apiBackend.currentUserCanEdit }}
+        <a
```
Comment (brylie) [inline client/views/api_backends/view_api_backend/details/details.html]: Wrap these buttons in a [Bootstrap button group](http://getbootstrap.com/components/#btn-groups), so they are aesthetically consistent with the ["My Managed APIs" template](https://github.com/apinf/api-umbrella-dashboard/blob/Issue%23616/client/views/api_backends/manage/manage.html#L8-L25).

File: client/views/api_backends/view_api_backend/details/details.js [hunk 23]
```diff
@@ -0,0 +1,8 @@
+Template.viewApiBackendDetails.events({
+  "click #deleteModal": function() {
+    var apiBackendId = this.apiBackend;
+    Modal.show('deleteApiBackendConfirmation', function() {
```
Comment (brylie) [inline client/views/api_backends/view_api_backend/details/details.js]: Add line comments here.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+    // Disable delete button to prevent multiple clicks
+    $('#deleteApi').prop("disabled", true);
+
+    // start spinner when calling API
+    instance.restCallStarted.set(true);
+
+    // REST call to Admin API for deletion from Umbrella
+    Meteor.call('deleteApiBackendOnApiUmbrella', apiUmbrellaApiId, function(error, apiUmbrellaWebResponse) {
+
+      if (apiUmbrellaWebResponse.http_status === 204) {
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Add line comment here, describing the purpose of this if statement in plain English.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+
+Template.deleteApiBackendConfirmation.helpers({
+  'savedBackendName': function() {
+
+    const instance = Template.instance();
+    return instance.backendName;
+  },
+
+  'restCallStarted': function() {
+    const instance = Template.instance();
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Add line comments here.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+  instance.backendId = instance.data._id;
+  instance.backendName = instance.data.name;
+  instance.restCallStarted = new ReactiveVar(false);
+});
+
+
+
+Template.deleteApiBackendConfirmation.helpers({
+  'savedBackendName': function() {
+
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Add line comment(s) here.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 27]
```diff
@@ -0,0 +1,69 @@
+Template.deleteApiBackendConfirmation.onCreated(function() {
+  // reference to this template
+  const instance = this;
+
+  // instance variables
+  instance.backendId = instance.data._id;
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Add line comments describing purpose of each variable in plain English.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js [hunk 28]
```diff
@@ -0,0 +1,69 @@
+Template.deleteApiBackendConfirmation.onCreated(function() {
+  // reference to this template
+  const instance = this;
+
+  // instance variables
```
Comment (brylie) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: Great variable names here. Go ahead and add line comments describing how they are used, so that a new developer would quickly grok their purpose.

File: client/views/api_backends/delete/deleteApiBackendConfirmation.js (hunk 21 above)
Comment (Alapan) [inline client/views/api_backends/delete/deleteApiBackendConfirmation.js]: @brylie I wrote the code for publishing in my last commit. I hope I did it correctly. Please review.

**Comment by brylie** [comment]
@Alapan, this looks great. I went ahead and fixed the Bootstrap button group markup, so that it renders correctly. Please review and merge this PR, if everything looks good.

**Comment by Alapan** [comment]
@brylie Merged.
